src/install/install.zig

```diff
@@ -4970,49 +4970,107 @@ pub const PackageManager = struct {
-                        Output.prettyErrorln("Error {s} while chdir - {s}", .{ @errorName(err), bun.span(chdir) });
+        const package_json_file = package_json_file_ orelse brk: {
+            var this_cwd = original_cwd;
+            const child_json = child: {
+                while (true) {
+                    var dir = std.fs.openDirAbsolute(this_cwd, .{}) catch |err| {
+                        Output.prettyErrorln("Error {s} accessing {s}", .{ @errorName(err), this_cwd });
                         Output.flush();
                         return err;
                     };
-
-                    break :brk std.fs.cwd().openFileZ("package.json", .{ .mode = .read_write }) catch {
-                        this_cwd = parent;
-                        continue :outer;
+                    defer dir.close();
+                    break :child dir.openFileZ("package.json", .{ .mode = .read_write }) catch {
+                        if (std.fs.path.dirname(this_cwd)) |parent| {
+                            this_cwd = parent;
+                            continue;
+                        } else {
+                            break;
+                        }
                     };
                 }
-
-                bun.copy(u8, &cwd_buf, original_cwd);
-                cwd_buf[original_cwd.len] = 0;
-                var real_cwd: [:0]u8 = cwd_buf[0..original_cwd.len :0];
-                std.os.chdirZ(real_cwd) catch {};
-
                 return error.MissingPackageJSON;
             };
-        }
+            // Check if this is a workspace; if so, use root package
+            const child_cwd = this_cwd;
+            var found = false;
+            while (std.fs.path.dirname(this_cwd)) |parent| {
+                var dir = std.fs.openDirAbsolute(parent, .{}) catch break;
+                defer dir.close();
+                const json_file = dir.openFileZ("package.json", .{ .mode = .read_write }) catch {
+                    this_cwd = parent;
+                    continue;
+                };
+                defer if (!found) json_file.close();
+                const json_stat = try json_file.stat();
+                const json_buf = try ctx.allocator.alloc(u8, json_stat.size + 64);
+                defer ctx.allocator.free(json_buf);
+                const json_len = try json_file.preadAll(json_buf, 0);
+                var path_buf: [bun.MAX_PATH_BYTES]u8 = undefined;
+                const json_path = try bun.getFdPath(json_file.handle, &path_buf);
+                const json_source = logger.Source.initPathString(
+                    json_path,
+                    json_buf[0..json_len],
+                );
+                initializeStore();
+                const json = try json_parser.ParseJSONUTF8(&json_source, ctx.log, ctx.allocator);
+                if (json.asProperty("workspaces")) |prop| {
+                    var workspace_names = bun.StringArrayHashMap(string).init(ctx.allocator);
```

Comment:
```zig
var workspace_names = bun.StringMap.init(ctx.allocator, true);
defer workspace_names.deinit();
```

Reply:
You mean that [one from six lines below](https://github.com/oven-sh/bun/pull/2834/files#diff-2d4db027e059280840de7dfe46dda1752467e66c1c3014369ed701b026cd06fcR5032)? 🤔

Reply:
no i meant we could remove the code for iterating through .values() and .keys() to free them and re-use an existing abstraction `StringMap` for this to do that for us

Reply:
Ah okay &minus; will upgrade the one within `Package.parseWithJSON()` too then.

Reply:
Done in https://github.com/oven-sh/bun/pull/2834/commits/28c24d8b5b412b57f7f00228c98dc46bdede16f7